fendermint/vm/interpreter/src/fvm/exec.rs

```diff
@@ -149,17 +152,22 @@ where
     async fn deliver(
         &self,
         mut state: Self::State,
-        msg: Self::Message,
+        mut msg: Self::Message,
     ) -> anyhow::Result<(Self::State, Self::DeliverOutput)> {
         let (apply_ret, emitters, latency) = if msg.from == system::SYSTEM_ACTOR_ADDR {
             let (execution_result, latency) = measure_time(|| state.execute_implicit(msg.clone()));
             let (apply_ret, emitters) = execution_result?;
 
             (apply_ret, emitters, latency)
         } else {
+            // TODO: maybe compare the gas limits is better?
+            msg.gas_limit = msg.gas_limit.min(self.gas.available_block_gas());
```

Comment:
This is incorrect here. Instead, we should validate that the sum of all transaction gas limits doesn't exceed the block limit, at block validation time `prepare_proposal`. That guarantees we have enough gas to process all transactions even if they squeeze their gas limits to the max. Then this situation cannot occur.